### PR TITLE
Clean up comments in nestkernel

### PR DESCRIPTION
**Pull request by jessica-mitchell**
This PR aims to make the comments (doxygen and c++) more consistent and useful.
So far, only a few files have been touched because I would like some feedback as to whether I'm going in the right direction before proceeding. @jougs could you take a look?

Changes implemented:

* C++ comments single line (max 2 lines if sentence is long): // 
* C++ comments multile line: /* * * * */
* doxygen comments single line (max 2 lines if sentence is long): //!
* doxygen comments multiline: /** * * * * */
* doxygen comments side: //!<
* avoid doxygen comments in .cpp files, move to header file 
* avoid duplicating code in comments; only include code(parameters, functions etc) that also have additional context,  remove redundancies
* include the variable name in functions in header file to match cpp file.

### CODE REVIEW

File: nestkernel/conn_builder.h [hunk 1]
```diff
@@ -105,7 +109,19 @@ class ConnBuilder
...
+ * Returns 0 if the target is either on another
+ * MPI machine or another thread. Returns 1 otherwise.
+ *
+ * @param snode_id id of the source
+ * @param tnode_id id of the target
+ * @param tid thread id
+ * @param update amount of connected synaptic elements to update
+ * @return
+ */
+  bool change_connected_synaptic_elements( index snode_id, index tnode_id, const int thread tid, int update );
```
Comment (jougs) [inline nestkernel/conn_builder.h:124]: This should fix the build problems:

```suggestion
  bool change_connected_synaptic_elements( index snode_id, index tnode_id, const int tid, int update );
```

**Review by jougs — COMMENTED** [review]
Thanks for making the chaos and invonsistencies visible. I'm reviewing in greater detail once the crying stops. For now, I have a couple of generic comments:

* I prefer the following doxygen style for functions and classes, even if the comment itself is only a single line:
  ```
  /**
   * ...
   */
  ```
  (I also think this is the predominant style presently used, so it will save you a lot of trouble down the line :smile:)
* I would use `//!` for long one-liners above variables and `//!<` for short ones behind them

Maybe we should set up a video session where we go through the [Doxygen documentation](https://www.doxygen.nl/manual/docblocks.html) and ask people what they (dis)like most?

File: nestkernel/clopath_archiving_node.cpp [hunk 2]
```diff
@@ -64,13 +64,15 @@ nest::ClopathArchivingNode::ClopathArchivingNode( const ClopathArchivingNode& n
...
-  delayed_u_bars_idx_ = 0;
+  /*
+   * Implementation of the delay of the convolved membrane potentials. This
+   * delay is not described in Clopath et al. 2010 but is present in the code
+   * (https://senselab.med.yale.edu/ModelDB/showmodel.cshtml?model=144566) on
+   * ModelDB which was presumably used to create the figures in the paper.
+   * Since we write into the buffer before we read from it, we have to
+   * add 1 to the size of the buffers.
+   */
+  layed_u_bars_idx_ = 0;
```
Comment (jougs) [inline nestkernel/clopath_archiving_node.cpp:75]: Here's another one for fixing the build
```suggestion
  delayed_u_bars_idx_ = 0;
```

**Comment by jessica-mitchell** [comment]
> Thanks for making the chaos and invonsistencies visible. I'm reviewing in greater detail once the crying stops. For now, I have a couple of generic comments:
> 
>     * I prefer the following doxygen style for functions and classes, even if the comment itself is only a single line:
>       ```
>       /**
>        * ...
>        */
>       ```
Sure, I think you're right that it's more common
> 
>     * I would use `//!` for long one-liners above variables and `//!<` for short ones behind them

When you say above variables, can you give me an example of the syntax you'd expect this kind of comment? Where would not want this comment style? (it's just easier if I have an example to look up, because it's easy to get confused staring at slashes and askterisks all day ;) )
> 
> Maybe we should set up a video session where we go through the [Doxygen documentation](https://www.doxygen.nl/manual/docblocks.html) and ask people what they (dis)like most?

Yah sure, I think that could work.

**Comment by jougs** [comment]
Here are some examples that hopefully clarify my style preferences mentioned above:
```c++
/**
 * Set the minimal and maximal delay and override automatically determined values.
 */
void set_delay_extrema(delay min_delay, delay, max_delay();

//! The variable min_delay holds the value for the smallest transmission delay in the network in steps
delay min_delay;

delay max_delay //!< The largest transmission delay in the network (steps)
```

In real code, I would then actually prefer if similar variables were documented consistently, i.e. both `min_delay` and `max_delay` with either the long version, or with the short one, depending on the length of the comment.

File: doc/devdoc/cppcomments.rst [hunk 3]
```diff
@@ -0,0 +1,40 @@
+Rules for C++  code comments
+============================
+
+* C++ comments single line (max 2 lines if sentence is long): //
+* C++ comments multile line: ``/* * * * */``
```
Comment (heplesser) [inline doc/devdoc/cppcomments.rst:5]: This looks a bit confusing. I'd typeset it over multiple lines.

File: doc/devdoc/cppcomments.rst [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+* avoid doxygen comments in .cpp files, move to header file
+* avoid duplicating code in comments; only include code(parameters, functions etc) that also have additional context, remove redundancies
+* include the variable name in functions in header file to match cpp file.
+
+
+Functions and classes
+---------------------
+
+doxygen style: ``/** * */``
```
Comment (heplesser) [inline doc/devdoc/cppcomments.rst:15]: I would typeset this over multiple lines as well.

File: doc/devdoc/cppcomments.rst [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+
+Variables
+---------
+
+
+doxygen comments single line (max 2 lines if sentence is long): //!
```
Comment (heplesser) [inline doc/devdoc/cppcomments.rst:22]: At the beginning of sentences, I'd capitalize in keeping with the doxygen website.
```suggestion
Doxygen comments single line (max 2 lines if sentence is long): //!
```

Should there be backticks around //!, and similar below?

File: doc/devdoc/cppcomments.rst [hunk 6]
```diff
@@ -0,0 +1,40 @@
+Rules for C++  code comments
+============================
+
```
Comment (heplesser) [inline doc/devdoc/cppcomments.rst:3]: Maybe add a link to doxygen documentation?

File: doc/devdoc/cppcomments.rst [hunk 7]
```diff
@@ -0,0 +1,40 @@
+Rules for C++  code comments
+============================
```
Comment (heplesser) [inline doc/devdoc/cppcomments.rst:2]:
```suggestion
Rules for C++ code comments
===========================
```

File: nestkernel/archiving_node.h [hunk 8]
```diff
@@ -50,44 +50,32 @@ class ArchivingNode : public StructuralPlasticityNode
 {
 public:
   /**
-   * \fn ArchivingNode()
-   * Constructor.
+   * Constructor
    */
```
Comment (heplesser) [inline nestkernel/archiving_node.h:54]: Trivial comments like this on constructor, copy constuctor or assignment operator provide no information. I would remove them completely.

File: nestkernel/archiving_node.h [hunk 9]
```diff
@@ -50,44 +50,32 @@ class ArchivingNode : public StructuralPlasticityNode
...
-   * \fn ArchivingNode()
-   * Copy Constructor.
+   * Copy Constructor
    */
   ArchivingNode( const ArchivingNode& );
 
   /**
-   * \fn double get_K_value(long t)
-   * return the Kminus (synaptic trace) value at t (in ms). When the trace is
+   * Return the Kminus (synaptic trace) value at t (in ms). When the trace is
```
Comment (heplesser) [inline nestkernel/archiving_node.h:63]: If I am not entirely mistaken, the same rules apply to doxygen comments as to Python docstrings: the first line should give a short summary of what a function does, followed by an empty line and then, if relevant, details. So here 
```c++
/**
 * Return the Kminus (synaptic trace) value at t (in ms). 
 * 
 * When the trace is ...
```

File: nestkernel/archiving_node.h [hunk 10]
```diff
@@ -130,26 +114,25 @@ class ArchivingNode : public StructuralPlasticityNode
...
   void clear_history();
 
-  // number of incoming connections from stdp connectors.
-  // needed to determine, if every incoming connection has
-  // read the spikehistory for a given point in time
+  /*
+   * number of incoming connections from stdp connectors.
+   *  needed to determine, if every incoming connection has
+   *  read the spikehistory for a given point in time
+   */
```
Comment (heplesser) [inline nestkernel/archiving_node.h:135]:
```suggestion
  /*
   * Number of incoming connections from stdp connectors.
   *
   * It is needed to determine if every incoming connection has
   * read the spike history for a given point in time.
   */
```

File: nestkernel/clopath_archiving_node.cpp [hunk 11]
```diff
@@ -170,9 +172,11 @@ nest::ClopathArchivingNode::get_LTP_history( double t1,
   else
   {
     std::deque< histentry_extended >::iterator runner = ltp_history_.begin();
-    // To have a well defined discretization of the integral, we make sure
-    // that we exclude the entry at t1 but include the one at t2 by subtracting
-    // a small number so that runner->t_ is never equal to t1 or t2.
+    /*
```
Comment (heplesser) [inline nestkernel/clopath_archiving_node.cpp:175]: I'd put an empty line above this comment.

**Review by heplesser — CHANGES_REQUESTED** [review]
@jessica-mitchell I added some comments to the guidelines as two examples and would like your and @jougs' feedback before commenting more.

**Comment by terhorstd** [comment]
@jessica-mitchell, do you have a status update on this one?

**Comment by jougs** [comment]
@jessica-mitchell: What is the status of this? Thanks!

File: doc/devdoc/cppcomments.rst (hunk 4 above)
Comment (jessica-mitchell) [reply doc/devdoc/cppcomments.rst]: done

File: doc/devdoc/cppcomments.rst (hunk 5 above)
Comment (jessica-mitchell) [reply doc/devdoc/cppcomments.rst]: yes

File: nestkernel/archiving_node.h (hunk 8 above)
Comment (jessica-mitchell) [reply nestkernel/archiving_node.h]: Removed these, also removed destructor comments as well

File: nestkernel/archiving_node.h (hunk 9 above)
Comment (jessica-mitchell) [reply nestkernel/archiving_node.h]: Added a new line after first line description in comments

File: doc/devdoc/cppcomments.rst (hunk 6 above)
Comment (jessica-mitchell) [reply doc/devdoc/cppcomments.rst]: done

File: doc/htmldoc/developer_space/cppcomments.rst [hunk 12]
```diff
@@ -0,0 +1,80 @@
+Guidelines for C++  code comments
```
Comment (jessica-mitchell) [inline doc/htmldoc/developer_space/cppcomments.rst:1]:
```suggestion
Guidelines for C++ code comments
```

File: nestkernel/conn_builder.h [hunk 13]
```diff
@@ -308,10 +353,42 @@ class AllToAllBuilder : public ConnBuilder
   }
 
 protected:
+<<<<<<< HEAD
```
Comment (jessica-mitchell) [inline nestkernel/conn_builder.h:356]:
```suggestion
```

File: doc/htmldoc/developer_space/cppcomments.rst [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+There are two types of code comments for C++ files: doxygen style and C++ style comments.
+
+* Doxygen styled comments are used for describing things like the purpose of the function, which parameters it accepts, and what output it generates.
+* Use Doxygen style comments in the header (``.h``)  files. Avoid using them in ``.cpp`` files.
+* Do not duplicate code in comments.
+..  Include the variable name in functions in header file to match cpp file.
+
+
+Generate HTML do doxygen comments
```
Comment (jessica-mitchell) [inline doc/htmldoc/developer_space/cppcomments.rst:12]:
```suggestion
Generate HTML from doxygen comments
```

File: doc/htmldoc/developer_space/cppcomments.rst [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (jessica-mitchell) [inline doc/htmldoc/developer_space/cppcomments.rst:13]:
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by jessica-mitchell** [comment]
@jougs @heplesser I think I addressed all the comments so far. In addition to removing the Constructor comments, I also removed Destructor.

I cleaned up the guidelines (doc/htmldoc/developer_space/cppcoments.rst) to be more clear, but I am unsure of how to describe when to use C++ comment style `/* */` versus doxygen style `/** */`.

File: nestkernel/connection.h [hunk 16]
```diff
@@ -319,22 +322,28 @@ Connection< targetidentifierT >::check_connection_( Node& dummy_target,
...
-  // 3. do the events sent by source mean the same thing as they are
-  // interpreted in target?
-  // note that we here use a bitwise and operation (&), because we interpret
-  // each bit in the signal type as a collection of individual flags
+  /*
+   * 3. do the events sent by source mean the same thing as they are
+   * interpreted in target?
+   * note that we here use a bitwise and operation (&), because we interpret
+   * each bit in the signal type as a collection of individual flags
+   */
```
Comment (jougs) [inline nestkernel/connection.h:346]: Revert

File: nestkernel/connection_id.h [hunk 17]
```diff
@@ -106,4 +106,4 @@ ConnectionID::get_port() const
 
 } // namespace
 
-#endif /* #ifndef CONNECTION_ID_H */
+#endif // #ifndef CONNECTION_ID_H
```
Comment (jougs) [inline nestkernel/connection_id.h:109]: Revert

**Review by jougs — COMMENTED** [review]
Here's a bunch of comments. Some of them are rather generic and should be applied in many places. Probably something like `git difftool --dir-diff upstream/master` could help with bulk reverting changes more easily than purely by hand. I will have another look once my comments are addressed.

File: nestkernel/common_properties_hom_w.h [hunk 18]
```diff
@@ -24,17 +24,17 @@
...
+ *  Name: static_synapse_hom_w - Static synapse type
+ *       using homogeneous weight, i.e. all synapses
+ *        feature the same w.
+ *
+ *  FirstVersion: April 2008
+ *
+ *  Author: Moritz Helias, Susanne Kunkel
+ *
+ *  SeeAlso: synapsedict, static_synapse
+ */
```
[the remark this answers is not in the document]
Comment (jessica-mitchell) [reply nestkernel/common_properties_hom_w.h:37]: I would propose that a separate PR be made to deal with any of  these SLI documentation sections that needs to be converted into reStructured Text

File: nestkernel/conn_builder_conngen.h [hunk 19]
```diff
@@ -61,6 +117,6 @@ class ConnectionGeneratorBuilder : public ConnBuilder
 
 } // namespace nest
 
-#endif /* ifdef HAVE_LIBNEUROSIM */
+#endif // ifdef HAVE_LIBNEUROSIM
 
-#endif /* ifdef CONN_BUILDER_CONNGEN_H */
+#endif // ifdef CONN_BUILDER_CONNGEN_H
```
[the remark this answers is not in the document]
Comment (jessica-mitchell) [reply nestkernel/conn_builder_conngen.h]: done

File: nestkernel/conn_builder_conngen.cpp [hunk 20]
```diff
@@ -349,4 +295,4 @@ ConnectionGeneratorBuilder::cg_get_ranges( RangeSet& ranges, const NodeCollectio
 
 } // namespace nest
 
-#endif /* ifdef HAVE_LIBNEUROSIM */
+#endif /* ifdef HAVE_LIBNEUROSIM  */ */
```
Comment (jessica-mitchell) [inline nestkernel/conn_builder_conngen.cpp:298]:
```suggestion
#endif /* ifdef HAVE_LIBNEUROSIM  */
```

File: nestkernel/deprecation_warning.cpp [hunk 21]
```diff
@@ -58,4 +58,4 @@ DeprecationWarning::deprecation_warning( std::string name, std::string new_name
 }
 }
 
-#endif /* DEPRECATION_WARNING_IMPL_H */
+#endif /* DEPRECATION_WARNING_IMPL_H  */ */
```
Comment (jessica-mitchell) [inline nestkernel/deprecation_warning.cpp:61]:
```suggestion
#endif /* DEPRECATION_WARNING_IMPL_H  */
```

File: nestkernel/mpi_manager.cpp [hunk 22]
```diff
@@ -95,7 +95,7 @@ nest::MPIManager::init_mpi( int*, char*** )
   send_displacements_secondary_events_in_int_per_rank_.resize( 1, 0 );
 }
 
-#else /* HAVE_MPI */
+#else // HAVE_MPI
```
Comment (jessica-mitchell) [inline nestkernel/mpi_manager.cpp:98]:
```suggestion
#else /* HAVE_MPI */
```

File: nestkernel/mpi_manager.cpp [hunk 23]
```diff
@@ -124,11 +124,11 @@ nest::MPIManager::init_mpi( int* argc, char** argv[] )
     kernel().music_manager.init_music( argc, argv );
     // get a communicator from MUSIC
     set_communicator( static_cast< MPI_Comm >( kernel().music_manager.communicator() ) );
-#else  /* #ifdef HAVE_MUSIC */
+#else  // #ifdef HAVE_MUSIC
```
Comment (jessica-mitchell) [inline nestkernel/mpi_manager.cpp:127]:
```suggestion
#else  /* #ifdef HAVE_MUSIC */
```

File: nestkernel/mpi_manager.cpp [hunk 24]
```diff
@@ -124,11 +124,11 @@ nest::MPIManager::init_mpi( int* argc, char** argv[] )
...
     int provided_thread_level;
     MPI_Init_thread( argc, argv, MPI_THREAD_FUNNELED, &provided_thread_level );
     set_communicator( MPI_COMM_WORLD );
-#endif /* #ifdef HAVE_MUSIC */
+#endif /* #ifdef HAVE_MUSIC/ */
```
Comment (jessica-mitchell) [inline nestkernel/mpi_manager.cpp:131]:
```suggestion
#endif /* #ifdef HAVE_MUSIC */
```

File: nestkernel/mpi_manager.cpp [hunk 25]
```diff
@@ -276,7 +276,7 @@ nest::MPIManager::mpi_finalize( int exitcode )
   }
 }
 
-#else /* #ifdef HAVE_MPI */
+#else // #ifdef HAVE_MPI
```
Comment (jessica-mitchell) [inline nestkernel/mpi_manager.cpp:279]:
```suggestion
#else /* #ifdef HAVE_MPI */
```

File: nestkernel/mpi_manager.cpp [hunk 26]
```diff
@@ -982,11 +978,8 @@ nest::MPIManager::time_communicate_alltoallv( int num_bytes, int samples )
   return foo.elapsed() / samples;
 }
 
-#else /* #ifdef HAVE_MPI */
+#else // #ifdef
```
Comment (jessica-mitchell) [inline nestkernel/mpi_manager.cpp:981]:
```suggestion
#else /* #ifdef */
```

File: nestkernel/mpi_manager.cpp [hunk 27]
```diff
@@ -1100,4 +1090,4 @@ nest::MPIManager::communicate_recv_counts_secondary_events()
   send_displacements_secondary_events_in_int_per_rank_[ 0 ] = 0;
 }
 
-#endif /* #ifdef HAVE_MPI */
+#endif /* #ifdef HAVE_MPI  */ */
```
Comment (jessica-mitchell) [inline nestkernel/mpi_manager.cpp:1093]:
```suggestion
#endif /* #ifdef HAVE_MPI  */
```

File: nestkernel/mpi_manager.h [hunk 28]
```diff
@@ -755,7 +768,7 @@ MPIManager::time_communicate_alltoallv( int, int )
   return 0.0;
 }
 
-#endif /* HAVE_MPI */
+#endif /* HAVE_MPI  */ */
```
Comment (jessica-mitchell) [inline nestkernel/mpi_manager.h:771]:
```suggestion
#endif /* HAVE_MPI  */
```

File: nestkernel/mpi_manager_impl.h [hunk 29]
```diff
@@ -46,20 +46,18 @@ nest::MPIManager::get_process_id_of_vp( const thread vp ) const
 // Variable to hold the MPI communicator to use.
 #ifdef HAVE_MUSIC
 extern MPI::Intracomm comm;
-#else  /* #ifdef HAVE_MUSIC */
+#else  // * #ifdef HAVE_MUSIC
```
Comment (jessica-mitchell) [inline nestkernel/mpi_manager_impl.h:49]:
```suggestion
#else  /* #ifdef HAVE_MUSIC */
```

File: nestkernel/music_manager.cpp [hunk 30]
```diff
@@ -97,7 +97,7 @@ MUSICManager::enter_runtime( double h_min_delay )
   }
 }
 
-#else /* #ifdef HAVE_MUSIC */
+#else // #ifdef HAVE_MUSIC
```
Comment (jessica-mitchell) [inline nestkernel/music_manager.cpp:100]:
```suggestion
#else /* #ifdef HAVE_MUSIC */
```

File: nestkernel/music_manager.cpp [hunk 31]
```diff
@@ -124,7 +124,7 @@ MUSICManager::music_finalize()
 
   music_runtime->finalize();
   delete music_runtime;
-#else /* #ifdef HAVE_MUSIC */
+#else // #ifdef HAVE_MUSIC
```
Comment (jessica-mitchell) [inline nestkernel/music_manager.cpp:127]:
```suggestion
#else /* #ifdef HAVE_MUSIC */
```

File: nestkernel/common_properties_hom_w.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (jougs) [reply nestkernel/common_properties_hom_w.h]: I would not bother with any SLI things anymore at this point as these comments will anyway just be deleted in PyNEST NG together with all of SLI.

**Comment by jessica-mitchell** [comment]
@jougs the conflict is fixed

**Comment by jessica-mitchell** [comment]
@jougs ping!

**Comment by jougs** [comment]
@jessica-mitchell: [pong!](https://github.com/jessica-mitchell/nest-simulator/pull/19)

**Review by jougs — APPROVED** [review]
Many thanks for this heroic effort and your patience. Merging without second review due to discussion in several meetings and because this "only" touches documentation.

**Pull request merged**